### PR TITLE
VaultParams refactoring

### PR DESCRIPTION
**Pull request by Krishnakumarskr**

### CODE REVIEW

File: packages/contracts/src/interface/ICredbull.sol [hunk 1]
```diff
@@ -27,6 +27,56 @@ interface ICredbull {
...
+    struct KycParams {
+        address kycProvider;
+        uint256 depositThresholdForWhitelisting;
+    }
+
+    struct MaxCapParams {
+        uint256 maxCap;
+    }
+
+    struct FixedYieldVaultParams {
```
Comment (jplodge-pro) [inline packages/contracts/src/interface/ICredbull.sol:62]: Are we not moving the specific param Structs to the associated contract?

**Review by jplodge-pro — COMMENTED** [review]
After our discussion, I'm not sure if I was supposed to mention the issues we discussed! So I did :D

File: packages/contracts/src/interface/ICredbull.sol (hunk 1 above)
Comment (jplodge-pro) [reply packages/contracts/src/interface/ICredbull.sol:62]: As discussed, this is already a large PR, so moving the 'params' structs will be another PR.

File: packages/api/src/modules/vaults/vaults.service.ts [hunk 2]
```diff
@@ -100,6 +105,54 @@ export class VaultsService {
...
+
+    const matureWindowParam = {
+      opensAt: params.redemptionOpensAt,
+      closesAt: params.redemptionClosesAt,
+    };
+
+    const windowVaultParams = {
+      depositWindow: depositWindowParam,
+      matureWindow: matureWindowParam,
+    };
```
Comment (lucasia) [inline packages/api/src/modules/vaults/vaults.service.ts:135]: for the deposit windows, i was thinking we could have a struct to represent the open and close at.  generally, i'd prefer object types that are more descriptive over simple primitives.  

for example below, this also raises the question, shouldn't a window be between two Dates (not numbers).

```    
interface WindowParam {
  opensAt: BigNumber;
  closesAt: BigNumber;
}
```


```    
    const depositWindowParam: WindowParam = {
      opensAt: params.depositOpensAt,
      closesAt: params.depositClosesAt,
    };

    const matureWindowParam: WindowParam = {
      opensAt: params.redemptionOpensAt,
      closesAt: params.redemptionClosesAt,
    };```

File: packages/contracts/src/CredbullFixedYieldVaultWithUpside.sol [hunk 3]
```diff
@@ -3,9 +3,10 @@ pragma solidity ^0.8.19;
 
 import { IERC20 } from "@openzeppelin/contracts/interfaces/IERC20.sol";
 import { UpsideVault } from "./vaults/UpsideVault.sol";
+import { ICredbull } from "./interface/ICredbull.sol";
 
 contract CredbullFixedYieldVaultWithUpside is UpsideVault {
-    constructor(VaultParams memory params, IERC20 _token, uint256 _collateralPercentage)
+    constructor(ICredbull.FixedYieldVaultParams memory params, IERC20 _token, uint256 _collateralPercentage)
```
Comment (lucasia) [inline packages/contracts/src/CredbullFixedYieldVaultWithUpside.sol:9]: suggest to group the token and collateral percentage here into a UpsideVaultParams struct.

1. suggest to rename token to something more explicit.  tokenForLock or similar.  having it within the UpsideVaultParams struct will also help with context.  we have so many tokens on Vaults, it isn't obvious what this is!

1. what is the collateralPercentage?  is this the upsideReward that we get?  or tokensRequiredForLock?  let's rename it as appropriate.

File: packages/contracts/src/base/CredbullBaseVault.sol [hunk 4]
```diff
@@ -51,16 +51,19 @@ abstract contract CredbullBaseVault is ICredbull, ERC4626, Pausable {
...
     /**
      *
-     * @param params - Vault parameters
+     * @param baseVaultParams - Base vault parameters
      */
-    constructor(VaultParams memory params) ERC4626(params.asset) ERC20(params.shareName, params.shareSymbol) {
-        if (params.custodian == address(0) || address(params.asset) == address(0)) {
+    constructor(BaseVaultParams memory baseVaultParams)
+        ERC4626(baseVaultParams.asset)
+        ERC20(baseVaultParams.shareName, baseVaultParams.shareSymbol)
```
Comment (lucasia) [inline packages/contracts/src/base/CredbullBaseVault.sol:58]: great, love it!  keeps our base very simple!

File: packages/contracts/src/interface/ICredbull.sol [hunk 5]
```diff
@@ -27,6 +27,56 @@ interface ICredbull {
...
+    }
+
+    struct MaturityVaultParams {
+        BaseVaultParams baseVaultParams;
+        ContractRoles contractRoles;
+        WindowVaultParams windowVaultParams;
+        KycParams kycParams;
+        MaxCapParams maxCapParams;
+        uint256 promisedYield;
+    }
```
Comment (lucasia) [inline packages/contracts/src/interface/ICredbull.sol:78]: duplicates FixedYieldVault - doesn't look like it's used anyhow.  remove it.

**Review by lucasia — COMMENTED** [review]
please see comments, looks good overall!  please incorporate those few suggested changes.

File: packages/api/src/modules/vaults/vaults.service.ts (hunk 2 above)
Comment (jplodge-pro) [reply packages/api/src/modules/vaults/vaults.service.ts:135]: I love the idea of `interface`s with the param values, then using multiple inheritance to realise a single params. 
Sadly, interfaces don't allow state variables [link](https://docs.soliditylang.org/en/v0.8.26/contracts.html#interfaces) (I would love to be wrong here!)
So we would have to use accessor functions, which would be a lot more cumbersome and would reduce the elegance of the solution I think.
```solidity
// Read-only interface
interface IContractRoles {
    function getOwner() external returns (address);
    function getOperator() external returns (address);
    function getCustodian() external returns (address);
}

struct VaultParameters is IContractRoles {
...
```
Even still, as far as I can tell, only contracts themselves can realise `interface`s, so making this design a non-starter.

**Pull request merged**